File: python/examples/counter.py
"""Ordinary cooperative command: bounded counting with durable continuation."""
import argparse
import time
from cedegrid import WorkerContext

p = argparse.ArgumentParser()
p.add_argument("--steps", type=int, default=20)
p.add_argument("--delay", type=float, default=0.1)
a = p.parse_args()
context = WorkerContext.from_env()
start = (context.resume or {}).get("metadata", {}).get("completed_steps", 0)
for step in range(start, a.steps):
    if context.draining():
        context.checkpoint({"completed_steps": step})
        raise SystemExit(75)
    time.sleep(a.delay)
context.complete({"completed_steps": a.steps})

File: npm/examples/counter.ts
import { WorkerContext, DrainRequested } from 'cedegrid';
async function main(): Promise<void> {
    const context = WorkerContext.fromEnv();
    let cursor = 0n;
    try {
        for (; cursor < 1000n; cursor++) {
            context.safePoint();
            if (cursor % 100n === 0n)
                await context.checkpoint({ cursor });
        }
        await context.complete({ cursor });
    }
    catch (error) {
        if (!(error instanceof DrainRequested))
            throw error;
        await context.checkpoint({ cursor });
    }
}
main().catch(error => { console.error(error); process.exitCode = 1; });
